graph/network/diffusion.go

```diff
@@ -0,0 +1,212 @@
+// license that can be found in the LICENSE file.
+
+package network
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/mat"
+)
+
+// Diffuse performs a heat diffusion across nodes of the undirected
+// graph described by the given Laplacian using the initial heat distribution,
+// h, according to the Laplacian with a diffusion time of t.
+// The resulting heat distribution is returned, written into the map dst and
+// returned,
+//  d = exp(-Lt)×h
+// where L is the graph Laplacian. Indexing into h and dst is defined by the
+// Laplacian Index field. If dst is nil, a new map is created.
+//
+// Nodes without corresponding entries in h are given an initial heat of zero,
+// and entries in h without a corresponding node in the original graph are
+// not altered when written to dst.
+func Diffuse(dst, h map[int64]float64, by Laplacian, t float64) map[int64]float64 {
+	heat := make([]float64, len(by.Index))
+	for id, i := range by.Index {
+		heat[i] = h[id]
+	}
+	v := mat.NewVecDense(len(heat), heat)
+
+	var m, tl mat.Dense
+	tl.Scale(-t, by)
+	m.Exp(&tl)
+	v.MulVec(&m, v)
+
+	if dst == nil {
+		dst = make(map[int64]float64)
+	}
+	for i, n := range v.RawVector().Data {
+		dst[by.Nodes[i].ID()] = n
+	}
+	return dst
+}
+
+// DiffuseToEquilibrium performs a heat diffusion across nodes of the
+// graph described by the given Laplacian using the initial heat
+// distribution, h, according to the Laplacian until the update function
+//  h_{n+1} = h_n - L×h_n
+// results in a 2-norm update difference within tol, or iter updates have
+// been made.
+// The resulting heat distribution is returned as eq, written into the map dst,
+// and a boolean indicating whether the equilibrium converged to within tol.
+// Indexing into h and dst is defined by the Laplacian Index field. If dst
+// is nil, a new map is created.
+//
+// Nodes without corresponding entries in h are given an initial heat of zero,
+// and entries in h without a corresponding node in the original graph are
+// not altered when written to dst.
+func DiffuseToEquilibrium(dst, h map[int64]float64, by Laplacian, tol float64, iters int) (eq map[int64]float64, ok bool) {
+	heat := make([]float64, len(by.Index))
+	for id, i := range by.Index {
+		heat[i] = h[id]
+	}
+	v := mat.NewVecDense(len(heat), heat)
+
+	last := make([]float64, len(by.Index))
+	for id, i := range by.Index {
+		heat[i] = h[id]
+	}
+	lastV := mat.NewVecDense(len(last), last)
+
+	var tmp mat.VecDense
+	for {
+		lastV, v = v, lastV
+		tmp.MulVec(by.Matrix, lastV)
+		v.SubVec(lastV, &tmp)
+		if normDiff(heat, last) < tol {
+			ok = true
+			break
+		}
+		iters--
+		if iters < 0 {
+			break
+		}
+	}
+
+	if dst == nil {
+		dst = make(map[int64]float64)
+	}
+	for i, n := range v.RawVector().Data {
+		dst[by.Nodes[i].ID()] = n
+	}
+	return dst, ok
+}
+
+// Laplacian is a graph Laplacian matrix.
+type Laplacian struct {
+	// Matrix holds the Laplacian matrix.
+	mat.Matrix
```

Comment:
A free flow of thoughts: if we consider only `NewLaplacian` and `NewSymNormLaplacian`, they take `graph.Undirected`, so both Laplacians are actually symmetric. There could be another type, say `TransitionMatrix`, that would be used for the random walks by operating on probability vectors while the Laplacian would be used to diffuse general scalar fields on a graph. Then the interface here could be `mat.Symmetric`. This would also give an important message about its eigenvalues which are an important point in all this. By having TransitionMatrix we would offer another matrix describing the graph (even though obtainable from the random walk Laplacian), and also the iterative update would be simpler. TransitionMatrix is not symmetric and could be generated also for directed graphs (as is done now). The damping would then not be specified in `NewTransitionMatrix` but in `RandomWalk(TransitionMatrix, damp float64)`. Finally, but not so importantly, the types could live in `graph`, so to the outside they would like as `graph.Laplacian`, `graph.TransitionMatrix`.

Reply:
I need to think about this. I know that L^{rw} is an abuse of the term Laplacian, but it's in use. Note that L^{rw} for undirected graphs is still not symmetric.